dj_cqrs/transport/rabbit_mq.py

```diff
@@ -92,48 +102,93 @@ def _produce(cls, payload):
-            dct['instance_data'],
-            dct.get('instance_pk'),
-            previous_data=dct.get('previous_data'),
-            correlation_id=dct.get('correlation_id'),
-        )
+        required_keys = {'instance_pk', 'signal_type', 'cqrs_id', 'instance_data'}
+        for key in required_keys:
+            if key not in dct:
+                msg = "CQRS couldn't proceed, %s isn't found in body: %s."
+                logger.error(msg, key, body)
+                ch.basic_reject(delivery_tag=method.delivery_tag, requeue=False)
+                return
 
+        payload = TransportPayload.from_message(dct)
         cls.log_consumed(payload)
 
-        instance = None
+        delivery_tag = method.delivery_tag
+        if payload.is_expired():
+            cls._add_to_dead_letter_queue(ch, payload)
+            cls._nack(ch, delivery_tag)
+            return
+
+        instance, exception = None, None
         try:
             instance = consumer.consume(payload)
-        except Exception:
-            logger.error('CQRS service exception', exc_info=True)
+        except Exception as e:
+            exception = e
+            logger.error("CQRS service exception", exc_info=True)
 
-        if instance:
-            ch.basic_ack(delivery_tag=method.delivery_tag)
-            cls.log_consumed_accepted(payload)
+        if instance and exception is None:
+            cls._ack(ch, delivery_tag, payload)
         else:
-            ch.basic_nack(delivery_tag=method.delivery_tag)
-            cls.log_consumed_denied(payload)
+            cls._fail_message(
+                ch, delivery_tag, payload, exception, delay_queue,
+            )
+
+    @classmethod
+    def _fail_message(cls, channel, delivery_tag, payload, exception, delay_queue):
+        cls.log_consumed_failed(payload)
+        model_cls = ReplicaRegistry.get_model_by_cqrs_id(payload.cqrs_id)
+        if model_cls is None:
+            logger.error("Model for cqrs_id {} is not found.".format(payload.cqrs_id))
+            cls._nack(channel, delivery_tag)
+            return
+
+        if model_cls.should_retry_cqrs(payload.retries, exception):
+            delay = model_cls.get_cqrs_retry_delay(payload.retries)
+            eta = timezone.now() + timedelta(seconds=delay)
+
+            delayed_message = DelayMessage(delivery_tag, payload, eta)
+            delay_queue.put(delayed_message)
+            cls.log_delayed(payload, delay, delayed_message.eta)
+        else:
+            cls._add_to_dead_letter_queue(channel, payload)
+            cls._nack(channel, delivery_tag)
+
+    @classmethod
+    def _add_to_dead_letter_queue(cls, channel, payload):
+        dead_message_ttl = settings.CQRS.get('dead_message_ttl', DEFAULT_DEAD_MESSAGE_TTL)
+        expiration = None
+        if dead_message_ttl is not None:
+            expiration = str(dead_message_ttl * 1000)  # milliseconds
+
+        exchange = cls._get_common_settings()[-1]
+        cls._produce_message(
```

Comment:
I guess, that we can catch exception here on message producing and in such case, worker will die. If I am right, please, fix that.

Reply:
We already catch Connection exceptions at consume main loop (while True) - it will reconnect and message will be retried (at this worker or another).

You wand to handle any exception and skip adding to Dead letter queue?